penaltymodel_lp/penaltymodel/lp/generation.py

```diff
@@ -0,0 +1,99 @@
+from bisect import bisect_left
+import dimod
+from itertools import product
+import numpy as np
+import penaltymodel.core as pm
+from scipy.optimize import linprog
+
+
+#TODO: would be nice to have a file for default linear energy ranges (currently, [-2, 2]); quad energy [-1, 1]
+def _get_lp_matrix(spin_states, nodes, edges, offset_weight, gap_weight):
+    # Set up an empty matrix
+    n_states = len(spin_states)
+    m_linear = len(nodes)
+    m_quadratic = len(edges)
+    matrix = np.empty((n_states, m_linear + m_quadratic + 2))   # +2 columns to account for offset and gap
+
+    # Populate linear terms (i.e. spin states)
+    matrix[:, :m_linear] = spin_states
+
+    # Populate quadratic terms
+    for j, (u, v) in enumerate(edges):
+        u_ind = bisect_left(nodes, u)
+        v_ind = bisect_left(nodes, v)
+        matrix[:, j + m_linear] = np.multiply(matrix[:, u_ind], matrix[:, v_ind])
+
+    # Populate offset and gap columns, respectively
+    matrix[:, -2] = offset_weight
+    matrix[:, -1] = gap_weight
+    return matrix
+
+
+def generate_bqm(graph, table, decision_variables,
+                 linear_energy_ranges=None, quadratic_energy_ranges=None):
+
+    # TODO: better way of saying that toy does not deal with auxiliary
+    # Check for auxiliary variables in the graph
+    if len(graph) != len(decision_variables):
+        raise ValueError('Penaltymodel-lp does not handle problems with auxiliary variables')
+
+    if not linear_energy_ranges:
+        linear_energy_ranges = {}
+
+    if not quadratic_energy_ranges:
+        quadratic_energy_ranges = {}
+
+    # Sort graph information
+    # Note: nodes' and edges' order determine the column order of the LP
+    nodes = sorted(decision_variables)
+    edges = graph.edges
+
+    # Set variable names for lengths
+    m_linear = len(nodes)                   # Number of linear biases
+    m_quadratic = len(edges)                # Number of quadratic biases
+    n_valid = len(table)                    # Number of valid spin combinations
+    n_invalid = 2**(m_linear) - n_valid     # Number of invalid spin combinations
+
+    # Determining valid and invalid spin states
+    invalid_table = set(product([-1, 1], repeat=m_linear)) - set(table.keys())
```

Comment:
`invalid_table = set(config for config in product([-1, 1], repeat=m_linear) if config not in table)` would be faster and use less memory